File: 백준/Gold/17404. RGB거리 2/RGB거리 2.py
import sys

input = sys.stdin.readline

n = int(input())

house = [list(map(int, input().split())) for _ in range(n)]
INF = int(1e6)

answer = INF

for start in range(3):
    dp = [[INF] * 3 for _ in range(n + 1)]
    dp[1][start] = house[0][start]
    
    for i in range(2, n):
        dp[i][0] = min(dp[i - 1][1], dp[i - 1][2]) + house[i - 1][0]
        dp[i][1] = min(dp[i - 1][0], dp[i - 1][2]) + house[i - 1][1]
        dp[i][2] = min(dp[i - 1][1], dp[i - 1][0]) + house[i - 1][2]

    for i in range(3):
        if i == start:
            continue
        
        answer = min(answer, min(dp[n - 1][(i + 1) % 3], dp[n - 1][(i + 2) % 3]) + house[-1][i])
        

print(answer)

File: 백준/Gold/17404. RGB거리 2/RGB거리 2.java
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        int[][] rgb = new int[n][3];
        for(int i=0 ; i<n ; i++){
            StringTokenizer tokenizer = new StringTokenizer(br.readLine());

            rgb[i][0] = Integer.parseInt(tokenizer.nextToken());
            rgb[i][1] = Integer.parseInt(tokenizer.nextToken());
            rgb[i][2] = Integer.parseInt(tokenizer.nextToken());
        }

        // 시작한 곳, 집 번호, 색
        int[][][] dp = new int[3][n + 1][3];

        int answer = Integer.MAX_VALUE;
        for(int start=0 ; start<3 ; start++){
            dp[start][2][start] = Integer.MAX_VALUE;
            dp[start][2][(start + 1) % 3] = rgb[0][start] + rgb[1][(start + 1) % 3];
            dp[start][2][(start + 2) % 3] = rgb[0][start] + rgb[1][(start + 2) % 3];

            for(int i=2 ; i<n-1 ; i++){
                dp[start][i + 1][0] = Math.min(dp[start][i][1], dp[start][i][2]) + rgb[i][0];
                dp[start][i + 1][1] = Math.min(dp[start][i][0], dp[start][i][2]) + rgb[i][1];
                dp[start][i + 1][2] = Math.min(dp[start][i][1], dp[start][i][0]) + rgb[i][2];
            }

            dp[start][n][(start + 1) % 3] = Math.min(dp[start][n - 1][start], dp[start][n - 1][(start + 2) % 3]) + rgb[n - 1][(start + 1) % 3];
            dp[start][n][(start + 2) % 3] = Math.min(dp[start][n - 1][start], dp[start][n - 1][(start + 1) % 3]) + rgb[n - 1][(start + 2) % 3];

            int minValue = Math.min(dp[start][n][(start + 1) % 3], dp[start][n][(start + 2) % 3]);
            answer = Math.min(minValue, answer);
        }

        System.out.println(answer);
        br.close();
    }
}